File: app/src/main/java/com/robifr/ledger/ui/selectcustomer/recycler/SelectCustomerListHolder.java
/**
 * Copyright (c) 2024 Robi
 *
 * Ledger is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Ledger is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Ledger. If not, see <https://www.gnu.org/licenses/>.
 */

package com.robifr.ledger.ui.selectcustomer.recycler;

import android.view.View;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.robifr.ledger.R;
import com.robifr.ledger.data.model.CustomerModel;
import com.robifr.ledger.databinding.CustomerCardWideBinding;
import com.robifr.ledger.ui.RecyclerViewHolder;
import com.robifr.ledger.ui.main.customer.CustomerCardWideNormalComponent;
import com.robifr.ledger.ui.selectcustomer.SelectCustomerFragment;
import java.util.Objects;

public class SelectCustomerListHolder extends RecyclerViewHolder<CustomerModel>
    implements View.OnClickListener {
  @NonNull private final SelectCustomerFragment _fragment;
  @NonNull private final CustomerCardWideBinding _cardBinding;
  @NonNull private final CustomerCardWideNormalComponent _normalCard;
  @Nullable private CustomerModel _boundCustomer;

  public SelectCustomerListHolder(
      @NonNull SelectCustomerFragment fragment, @NonNull CustomerCardWideBinding binding) {
    super(binding.getRoot());
    this._fragment = Objects.requireNonNull(fragment);
    this._cardBinding = Objects.requireNonNull(binding);
    this._normalCard =
        new CustomerCardWideNormalComponent(
            this._fragment.requireContext(), this._cardBinding.normalCard);

    this._cardBinding.cardView.setOnClickListener(this);
    // Don't set to `View.GONE` as the position will be occupied by checkbox.
    this._cardBinding.normalCard.menuButton.setVisibility(View.INVISIBLE);
  }

  @Override
  public void bind(@NonNull CustomerModel customer) {
    this._boundCustomer = Objects.requireNonNull(customer);
    final CustomerModel selectedCustomer =
        this._fragment.selectCustomerViewModel().initialSelectedCustomer();
    final boolean shouldChecked =
        selectedCustomer != null
            && selectedCustomer.id() != null
            && selectedCustomer.id().equals(this._boundCustomer.id());

    this._normalCard.setCustomer(this._boundCustomer);
    this._cardBinding.cardView.setChecked(shouldChecked);
  }

  @Override
  public void onClick(@NonNull View view) {
    Objects.requireNonNull(view);
    Objects.requireNonNull(this._boundCustomer);

    switch (view.getId()) {
      case R.id.cardView ->
          this._fragment.selectCustomerViewModel().onCustomerSelected(this._boundCustomer);
    }
  }
}

File: app/src/main/java/com/robifr/ledger/ui/selectcustomer/recycler/SelectCustomerListHolder.kt
/**
 * Copyright 2024 Robi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.robifr.ledger.ui.selectcustomer.recycler

import android.view.View
import androidx.core.view.isInvisible
import androidx.core.view.isVisible
import com.robifr.ledger.R
import com.robifr.ledger.components.CustomerCardWideComponent
import com.robifr.ledger.data.model.CustomerModel
import com.robifr.ledger.databinding.CustomerCardWideBinding
import com.robifr.ledger.ui.common.RecyclerViewHolder

class SelectCustomerListHolder(
    private val _cardBinding: CustomerCardWideBinding,
    private val _initialSelectedCustomerIds: () -> List<Long>,
    private val _customers: () -> List<CustomerModel>,
    private val _onCustomerSelected: (CustomerModel) -> Unit,
    private val _expandedCustomerIndex: () -> Int,
    private val _onExpandedCustomerIndexChanged: (Int) -> Unit
) : RecyclerViewHolder(_cardBinding.root), View.OnClickListener {
  private var _customerIndex: Int = -1
  private val _card: CustomerCardWideComponent =
      CustomerCardWideComponent(itemView.context, _cardBinding)

  init {
    _cardBinding.cardView.setOnClickListener { _onCustomerSelected(_customers()[_customerIndex]) }
    // Don't set menu button to gone as the position will be occupied by expand button.
    _cardBinding.normalCard.menuButton.isInvisible = true
    _cardBinding.normalCard.expandButton.isVisible = true
    _cardBinding.normalCard.expandButton.setOnClickListener(this)
    _cardBinding.expandedCard.menuButton.isInvisible = true
    _cardBinding.expandedCard.expandButton.isVisible = true
    _cardBinding.expandedCard.expandButton.setOnClickListener(this)
  }

  override fun bind(itemIndex: Int) {
    _customerIndex = itemIndex
    _card.reset()
    _card.setNormalCardCustomer(_customers()[_customerIndex])
    _card.setCardChecked(_initialSelectedCustomerIds().contains(_customers()[_customerIndex].id))
    setCardExpanded(
        _expandedCustomerIndex() != -1 &&
            _customers()[_customerIndex] == _customers()[_expandedCustomerIndex()])
  }

  override fun onClick(view: View?) {
    when (view?.id) {
      R.id.expandButton -> _onExpandedCustomerIndexChanged(_customerIndex)
    }
  }

  fun setCardExpanded(isExpanded: Boolean) {
    _card.setCardExpanded(isExpanded)
    // Only fill the view when it's shown on screen.
    if (isExpanded) _card.setExpandedCardCustomer(_customers()[_customerIndex])
  }
}
